python/ray/dag/compiled_dag_node.py

```diff
@@ -150,53 +186,60 @@ def _wrap_exception(exc):
     return wrapped
 
 
-def _exec_task(self, task: "ExecutableTask", idx: int) -> bool:
+def _exec_operation(self, task: "ExecutableTask", operation: DAGNodeOperation) -> bool:
     """
-    Execute the task.
+    Execute the `operation` which belongs to `task`.
     Args:
         task: The task to execute.
-        idx: The index of the task in the list of tasks of the actor.
+        operation: The operation to execute.
     Returns:
-        True if we are done executing all tasks of this actor, False otherwise.
+        True if we are done executing all operations of this actor, False otherwise.
     """
-    # TODO: for cases where output is passed as input to a task on
-    # the same actor, introduce a "IntraProcessChannel" to avoid the overhead
-    # of serialization/deserialization and synchronization.
-    method = getattr(self, task.method_name)
+    idx = operation.idx
+    op_type = operation.type
+
     input_reader = self._input_readers[idx]
     output_writer = self._output_writers[idx]
-    res = None
-    try:
-        res = input_reader.read()
-    except RayChannelError:
-        # Channel closed. Exit the loop.
-        return True
-
-    try:
-        _process_return_vals(res, return_single_output=False)
-    except Exception as exc:
-        # Previous task raised an application-level exception.
-        # Propagate it and skip the actual task. We don't need to wrap the
-        # exception in a RayTaskError here because it has already been wrapped
-        # by the previous task.
-        output_writer.write(exc)
-        return False
-
-    resolved_inputs = []
-    for task_input in task.task_inputs:
-        resolved_inputs.append(task_input.resolve(res))
-
-    try:
-        output_val = method(*resolved_inputs, **task.resolved_kwargs)
-    except Exception as exc:
-        output_val = _wrap_exception(exc)
-
-    try:
-        output_writer.write(output_val)
-    except RayChannelError:
-        # Channel closed. Exit the loop.
-        return True
-
+    ctx = ChannelContext.get_current().serialization_context
+
+    if op_type == DAGNodeOperationType.READ:
+        try:
```

Comment:
Let's separate these out into helper functions for readability.

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/ec43c663c9ed0ca0e85c818679bf61c5f28447f2

Reply:
I will add comments later